File: 20.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stk = []
  s.each_char do |c|
    if is_close?(c)
      return false if stk.empty?
      z = stk.pop
      return false if get_pair(c) != z
    else
      stk.push(c)
    end
  end
  return false if stk.empty?.!
  true
end

def is_close?(c)
  (c == ')' || c == ']' || c == '}')
end

def get_pair(c)
  r = ''
  r = '(' if c == ')'
  r = '[' if c == ']'
  r = '{' if c == '}'
  return r
end

File: 20.cpp
class Solution {
public:
  bool isValid(string s) {
    stack<char> stk;
    for(auto c:s) {
      switch(c) {
        char z;
        case ')':
          if(stk.empty()) return false;
          z = stk.top();
          stk.pop();
          if(z == '(') continue;
          return false;
        break;

        case '}':
          if(stk.empty()) return false;
          z = stk.top();
          stk.pop();
          if(z == '{') continue;
          return false;
        break;

        case ']':
          if(stk.empty()) return false;
          z = stk.top();
          stk.pop();
          if(z == '[') continue;
          return false;
        break;
        default:
          stk.push(c);
      }
    }
    if (stk.empty()) return true;
    return false;
  }
};
